benchmark/Benchmark.h

```diff
@@ -0,0 +1,172 @@
+     *  Most of the time a lambda, that calls the actual function to benchmark
+     *  with the needed parameters.
+     */
+    void addToExistingGroup(const std::string& groupDescriptor,
+        const std::string descriptor,
+        const std::function<void()>& functionToMeasure);
+
+    /*
+     * @brief Returns a const view of all the groups.
+     */
+    const ad_utility::HashMap<std::string, RecordGroup>& getGroups() const;
+
+    /*
+     * @brief Creates an empty table, which can be accesed/identified using the
+     *  descriptor.
+     *
+     * @param descriptor The name/identifier of the table.
+     * @param rowNames,columnNames The names for the rows/columns.
+     */
+    void addTable(const std::string descriptor,
+        const std::vector<std::string> rowNames,
+        const std::vector<std::string> columnNames);
+
+    /*
+     * @brief Measures the time needed for the execution of the given function and
+     * saves it as an entry in the table.
+     *
+     * @param tableDescriptor The identification of the table.
+     * @param row, column Where in the tables to write the measured time.
+     *  Starts with (0,0).
+     * @param functionToMeasure The function, that represents the benchmark.
+     *  Most of the time a lambda, that calls the actual function to benchmark
+     *  with the needed parameters.
+     */
+    void addToExistingTable(const std::string& tableDescriptor,
+        const size_t row, const size_t column,
+        const std::function<void()>& functionToMeasure);
+
+    /*
+     * @brief Returns a const view of all the tables.
+     */
+    const ad_utility::HashMap<std::string, RecordTable>& getTables() const;
+};
+
+/*
+ * This class exists as a round about way, of registering functions as
+ * benchmarks before entering the main function.
+ */
+class BenchmarkRegister {
+ 
+    // All registerd benchmark functions. That is, the functions, that do the
+    // setup needed for a benchmark and then record the execution time with
+    // the passed BenchmarkRecords.
+    // The calling of the registerd benchmark functions and the processing of
+    // the recorded times, will be done in a main function later.
+    static std::vector<std::function<void(BenchmarkRecords*)>> _registerdBenchmarks;
+
+  public:
+
+    /*
+     * @brief Register one, or more, functions as benchmarks by creating a
+     * global instance of this class. Shouldn't take up much space and I
+     * couldn't find a better way of doing it.
+     *
+     * @param benchmarks The functions can be passed as
+     *  `{&functionName1, &functionname2, ...}`. For more information about
+     *  their usage see _registerdBenchmarks . 
+     */
+    BenchmarkRegister(const std::vector<std::function<void(BenchmarkRecords*)>>& benchmarks);
+
+    // Return a const view of all registerd benchmarks.
+    static const std::vector<std::function<void(BenchmarkRecords*)>>& getRegisterdBenchmarks();
+};
```

Comment:
It is `registered` (with the second *e* ) (multiple occurences).

Reply:
Should be fixed, see [0042c16](https://github.com/ad-freiburg/qlever/pull/860/commits/0042c16f84926d2b77784e6b52c8dd05170a86f6).